fs: runtime deprecate closing FileHandle on garbage collection

jasnell opened the pull request:
Closing the FileHandle on garbage collection is a bad practice.
Runtime deprecate and indicate that an error will be thrown in
the future.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

vsemozhetbyt commented on doc/api/deprecations.md:
```diff
@@ -2533,6 +2566,7 @@ Passing a callback to [`worker.terminate()`][] is deprecated. Use the returned
 [`domain`]: domain.html
 [`ecdh.setPublicKey()`]: crypto.html#crypto_ecdh_setpublickey_publickey_encoding
 [`emitter.listenerCount(eventName)`]: events.html#events_emitter_listenercount_eventname
+[`fs.FileHandle`]: fs.html@fs_class_filehandle
```
```suggestion
[`fs.FileHandle`]: fs.html#fs_class_filehandle
```

cjihrig commented on doc/api/deprecations.md:
```diff
@@ -2498,6 +2498,39 @@ Type: Runtime
+<!-- YAML
+changes:
+  - version: REPLACEME
+    pr-url: REPLACEME
```
https://github.com/nodejs/node/pull/28396

BridgeAR replied:
```suggestion
    pr-url: https://github.com/nodejs/node/pull/28396
```

addaleax commented:
Hm, no strong opinion here, but using a deprecation warning means it’s only shown once rather than for each fd, right? Is that a good thing or should the separate warning be kept (e.g. the way it’s done for unhandled rejections)?

Trott commented on doc/api/deprecations.md:
```diff
@@ -2498,6 +2498,39 @@ Type: Runtime
+Please ensure that all `fs.FileHandle` objects are explicitly closed using
+`FileHandle.prototype.close()` when the `fs.FileHandle` is no longer needed.
+
+For instance,
```
Nit: Remove this line and end the previous sentence with `:`.

jasnell commented:
@addaleax... Sure, I could keep the other warning.

jasnell commented:
@addaleax ... added back the other warning, fixed a couple linting issues, and updated the test. PTAL :-)

@Trott @lpinca @cjihrig @tniessen ... may I ask y'all to take another look as well?

Trott commented on doc/api/deprecations.md:
```diff
@@ -2498,6 +2498,39 @@ Type: Runtime
+Type: Runtime
+
+Allowing a [`fs.FileHandle`][] object to be closed on garbage collection is
+deprecated. In the future, doing so will result in a thrown error that will
```
Optional nit: Since we can't be 100% sure about the future, perhaps `s/will/may/`?

jasnell replied:
```suggestion
deprecated. In the future, doing so will result in a thrown error that may
```

jasnell commented:
FYI... I'll be getting back to this very soon.

tniessen commented on test/parallel/test-fs-promises-file-handle-close.js:
```diff
@@ -0,0 +1,39 @@
+// be eligible for garbage collection.
+{
+  doOpen().then(common.mustCall((fd) => {
+    assert(fd);
```
Nit: Personally, I am not a fan of using `assert()` for anything other than booleans because it makes it harder for the reader to understand what the assertion is supposed to prevent: Does it prevent `fd === undefined`, `fd === false`, `fd === null`, or `fd === 0`? If all of these, I would use `assert(fd > 0)`.

Trott replied:
Maybe this?:

```suggestion
assert.strictEqual(typeof fd, 'number');
```

Trott replied:
Or this?:

```suggestion
    assert.ok(fd > 2); // fd 2 is stderr; open should return 3 or higher
```

jasnell commented on test/parallel/test-fs-promises-file-handle-close.js:
```diff
@@ -0,0 +1,39 @@
+'use strict';
+
+// Test that a runtime warning is emitted when a FileHandle object
+// is allowed to close on garbage collection. In this future, this
```
```suggestion
// is allowed to close on garbage collection. In the future, this
```

jasnell commented on doc/api/deprecations.md:
```diff
@@ -2498,6 +2498,39 @@ Type: Runtime
+Please ensure that all `fs.FileHandle` objects are explicitly closed using
+`FileHandle.prototype.close()` when the `fs.FileHandle` is no longer needed.
+
+For instance,
```
```suggestion
```

jasnell commented on doc/api/deprecations.md:
```diff
@@ -2498,6 +2498,39 @@ Type: Runtime
+terminate the process.
+
+Please ensure that all `fs.FileHandle` objects are explicitly closed using
+`FileHandle.prototype.close()` when the `fs.FileHandle` is no longer needed.
```
```suggestion
`FileHandle.prototype.close()` when the `fs.FileHandle` is no longer needed:
```

fhinkel commented:
ping @jasnell

jasnell commented:
Yep, haven't forgotten. Will get to this next week

On Thu, Oct 31, 2019, 17:43 F. Hinkelmann <notifications@github.com> wrote:

> ping @jasnell <https://github.com/jasnell>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/28396?email_source=notifications&email_token=AADLM6OCAR7TGMXG3APJTKDQRN3RLA5CNFSM4H2ZNPEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECZVGVQ#issuecomment-548623190>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADLM6KOFXFKD3HBG57OE43QRN3RLANCNFSM4H2ZNPEA>
> .
>

jasnell commented:
Ping @addaleax @Trott @lpinca @cjihrig @tniessen ... please take another look. Updated this to avoid using process state

jasnell commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/2259/ (link may not be immediately activated)

jasnell commented:
CITGM results show no relevant failures.

jasnell commented:
Landed in #28396

Uko commented:
Pardon my curiosity, but why "Closing the FileHandle on garbage collection is a bad practice?" I tried to google that up, but maybe smart people here can give me some pointers :)

tniessen commented:
@Uko The garbage collector should really only free memory that is not used anymore. Relying on the GC for cleaning up resources (e.g., closing file descriptors) is generally considered bad practice, because

- in many implementations, when and how the GC runs is not deterministic,
- resource cleanup might be slow and might therefore cause long GC times, and
- resource cleanup itself might produce errors, which cannot be handled by the GC.
